### ISSUE TITLE
Support indoor levels

### ISSUE THREAD

**Issue by brawer** [maplibre/maplibre-gl-native#703]
Could the MapLibre APIs support indoor levels? After a call like `setLevel(-2)`, the map would switch to the second underground level, similar to [OpenLevelUp](https://openlevelup.net/?l=-2#19/47.37800/8.53961) (see also [wiki](https://wiki.openstreetmap.org/wiki/OpenLevelUp)). As of January 2023, OSM contains ~1.2M objects with a `level` attribute (as per [taginfo](https://taginfo.openstreetmap.org/keys/level)), but most renderers only display features on the ground floor. Rendering indoor structures would be useful for shopping malls, airports, transit hubs, schools, hotels, etc.

In terms of implementation, I would recommend baking the needed information into the base vector tiles; this would give better latency than fetching non-default levels on demand. (As illustrated by the current OpenLevelUp site, whose latency is rather terrible). Also, in typical indoor use cases like airports or big buildings, there is often no cell phone reception, which is another reason for baking indoor data into the base vector tiles.

**Comment by louwers** [comment]
@brawer Are you specifically interested in this feature in MapLibre GL **Native**? Or MapLibre GL JS (for the web)? In case of the latter, kindly open an issue at the [maplibre-gl-js](https://github.com/maplibre/maplibre-gl-js) repository.

**Comment by wipfli** [comment]
Maybe indoor levels can already be handled by maplibre. Here is an example from the swiss train app by SBB:

![Screenshot_20230118-194926_SBB Mobile](https://user-images.githubusercontent.com/53421382/213270148-d7480f17-04b3-4f03-b3b1-ae89e014ef49.jpg)

**Comment by brawer** [comment]
What MapLibre API function should one call to switch the level? I couldn’t find this in the API docs.

**Comment by sebisteiner** [comment]
There is no API function to switch the level AFAIK.

We have defined filters in the layers of our style. For example:

    ["==", ["case", ["has", "level"], ["get", "level"], 0], 0]

Like this we only show features where the `level` property has the value 0 or is missing.

On level switch we adjust this filter. (We replace 0 with the new target level)

**Comment by brawer** [comment]
@sebisteiner Curious, how do you populate your level picker widget? Did you extend MapLibre's vector tiles handling to extract the set of available levels from the current viewport’s tile data? Or did you implement a separate RPC, independent of tiles, to fetch the available levels by bounding box?

**Comment by sebisteiner** [comment]
@brawer Yes, we fetch the available levels by bounding box. We have a separate GeoJSON source in our style where every feature represents a train station for which we have level informations. (Less than 50 stations, so this works quite well)

https://github.com/sbb-design-systems/sbb-angular/blob/main/src/journey-maps/angular/components/level-switch/services/query-map-features.ts#L12
